File: native/macos/Sources/MightyCore/PaneLayout.swift
import Foundation

/// A workspace's split tree. Leaves contain ordered tabs; split nodes contain
/// exactly two children after normalization. The wire schema is shared with Windows.
public struct PaneLayoutNode: Codable, Sendable, Equatable, Identifiable {
    public var id: String
    public var kind: String
    public var sessionIds: [String]
    public var selectedSessionId: String?
    public var axis: String?
    public var ratio: Double
    public var children: [PaneLayoutNode]

    public init(id: String = UUID().uuidString, kind: String, sessionIds: [String] = [], selectedSessionId: String? = nil, axis: String? = nil, ratio: Double = 0.5, children: [PaneLayoutNode] = []) {
        self.id = id; self.kind = kind; self.sessionIds = sessionIds; self.selectedSessionId = selectedSessionId
        self.axis = axis; self.ratio = ratio; self.children = children
    }

    enum CodingKeys: String, CodingKey { case id, kind, sessionIds, selectedSessionId, axis, ratio, children }
    public init(from decoder: Decoder) throws {
        let path = decoder.codingPath
        let depth = path.indices.dropFirst().filter { path[$0].intValue != nil && path[$0 - 1].stringValue == "children" }.count
        guard depth < PaneLayouts.maximumDepth else {
            throw DecodingError.dataCorrupted(.init(codingPath: path, debugDescription: "Pane layout is too deeply nested."))
        }
        let c = try decoder.container(keyedBy: CodingKeys.self)
        id = try c.decode(String.self, forKey: .id); kind = try c.decode(String.self, forKey: .kind)
        sessionIds = try c.decodeIfPresent([String].self, forKey: .sessionIds) ?? []
        selectedSessionId = try c.decodeIfPresent(String.self, forKey: .selectedSessionId)
        axis = try c.decodeIfPresent(String.self, forKey: .axis)
        ratio = try c.decodeIfPresent(Double.self, forKey: .ratio) ?? 0.5
        children = try c.decodeIfPresent([PaneLayoutNode].self, forKey: .children) ?? []
    }
}

public enum PaneLayouts {
    public static let maximumSessions = 128
    public static let maximumNodes = 255
    public static let maximumDepth = 16
    public static let minimumRatio = 0.15
    public static let maximumRatio = 0.85
    public static let viewModes = ["grid", "columns", "tabs", "custom", "focus"]

    /// One-time migration of the old global display mode. Focus applies only
    /// to the workspace that was active; subsequent workspaces never inherit it.
    public static func workspaceModes(workspaceIds: [String], activeWorkspaceId: String?, legacyMode: String, layouts: [String: PaneLayoutNode]?, savedModes: [String: String]?) -> [String: String] {
        var result: [String: String] = [:]
        for id in workspaceIds.prefix(64) where CoreValidation.identifier(id) {
            if let mode = savedModes?[id], viewModes.contains(mode) { result[id] = mode; continue }
            if savedModes == nil, legacyMode == "focus", id == activeWorkspaceId { result[id] = "focus"; continue }
            if let root = layouts?[id] { result[id] = root.kind == "tabs" ? "tabs" : "custom" }
            else if savedModes == nil, ["grid", "columns", "tabs"].contains(legacyMode) { result[id] = legacyMode }
            else { result[id] = "tabs" }
        }
        return result
    }

    public static func firstSelectedSession(in root: PaneLayoutNode?) -> String? {
        guard let root else { return nil }
        return find(root, depth: 0, where: { $0.kind == "tabs" && !$0.sessionIds.isEmpty }).flatMap { node in
            node.selectedSessionId.flatMap { node.sessionIds.contains($0) ? $0 : nil } ?? node.sessionIds.first
        }
    }

    public static func preset(sessionIds: [String], activeId: String? = nil, mode: String) -> PaneLayoutNode? {
        let ids = validSessionIds(sessionIds)
        guard !ids.isEmpty else { return nil }
        if mode == "tabs" || mode == "focus" || mode == "custom" {
            return PaneLayoutNode(kind: "tabs", sessionIds: ids, selectedSessionId: activeId.flatMap { ids.contains($0) ? $0 : nil } ?? ids[0])
        }
        func build(_ slice: ArraySlice<String>, depth: Int) -> PaneLayoutNode {
            if slice.count == 1 { return PaneLayoutNode(kind: "tabs", sessionIds: Array(slice), selectedSessionId: slice.first) }
            let middle = slice.startIndex + (slice.count + 1) / 2
            let left = slice[..<middle], right = slice[middle...]
            return PaneLayoutNode(kind: "split", axis: mode == "columns" || depth.isMultiple(of: 2) ? "horizontal" : "vertical",
                                  ratio: clamp(Double(left.count) / Double(slice.count)), children: [build(left, depth: depth + 1), build(right, depth: depth + 1)])
        }
        return build(ids[...], depth: 0)
    }

    /// Reconciles a persisted tree with this workspace's current sessions. New
    /// sessions join the first group; UI may then move them to its previous focus.
    public static func normalized(root: PaneLayoutNode?, sessionIds: [String], activeId: String? = nil) -> PaneLayoutNode? {
        let ids = validSessionIds(sessionIds)
        guard !ids.isEmpty else { return nil }
        let allowed = Set(ids)
        var seenSessions = Set<String>(), seenNodes = Set<String>(), count = 0
        func visit(_ source: PaneLayoutNode, depth: Int) -> PaneLayoutNode? {
            guard depth < maximumDepth, count < maximumNodes else { return nil }
            count += 1
            var id = source.id
            if !CoreValidation.identifier(id) || !seenNodes.insert(id).inserted { id = freshId(excluding: &seenNodes) }
            if source.kind == "tabs" {
                let tabs = source.sessionIds.prefix(maximumSessions).filter { allowed.contains($0) && seenSessions.insert($0).inserted }
                guard !tabs.isEmpty else { return nil }
                let selected = source.selectedSessionId.flatMap { tabs.contains($0) ? $0 : nil } ?? tabs[0]
                return PaneLayoutNode(id: id, kind: "tabs", sessionIds: tabs, selectedSessionId: selected)
            }
            guard source.kind == "split" else { return nil }
            let children = source.children.prefix(2).compactMap { visit($0, depth: depth + 1) }
            if children.count == 1 { return children[0] }
            guard children.count == 2 else { return nil }
            return PaneLayoutNode(id: id, kind: "split", axis: source.axis == "vertical" ? "vertical" : "horizontal", ratio: clamp(source.ratio), children: children)
        }
        var result = root.flatMap { visit($0, depth: 0) }
        let missing = ids.filter { !seenSessions.contains($0) }
        if result == nil { result = PaneLayoutNode(id: freshId(excluding: &seenNodes), kind: "tabs", sessionIds: ids, selectedSessionId: ids[0]) }
        else if !missing.isEmpty {
            var appended = false
            result = map(result!, depth: 0) { node in
                guard node.kind == "tabs", !appended else { return node }
                appended = true; var copy = node; copy.sessionIds.append(contentsOf: missing); return copy
            }
        }
        if let activeId, allowed.contains(activeId) { result = selecting(root: result, id: activeId) }
        return result
    }

    public static func selecting(root: PaneLayoutNode?, id: String) -> PaneLayoutNode? {
        guard let root, CoreValidation.identifier(id) else { return root }
        return map(root, depth: 0) { node in
            guard node.kind == "tabs", node.sessionIds.contains(id) else { return node }
            var copy = node; copy.selectedSessionId = id; return copy
        }
    }

    /// Add a new tab directly to its intended group. A stale target or a split
    /// beyond the depth limit changes nothing, including existing ratios/IDs.
    public static func inserting(root: PaneLayoutNode?, sessionId: String, targetGroupId: String? = nil, placement: String = "tab") -> PaneLayoutNode? {
        guard CoreValidation.identifier(sessionId), ["tab", "left", "right", "top", "bottom"].contains(placement) else { return root }
        guard let root else {
            guard targetGroupId == nil, placement == "tab" else { return nil }
            return PaneLayoutNode(kind: "tabs", sessionIds: [sessionId], selectedSessionId: sessionId)
        }
        let ids = allSessionIds(root)
        guard ids.count < maximumSessions, !ids.contains(sessionId), withinLimits(root),
              let target = find(root, depth: 0, where: { $0.kind == "tabs" && (targetGroupId == nil || $0.id == targetGroupId) }) else { return root }
        var nodeIds = Set(allNodeIds(root))
        let result = map(root, depth: 0) { node in
            guard node.id == target.id, node.kind == "tabs" else { return node }
            if placement == "tab" {
                var copy = node; copy.sessionIds.append(sessionId); copy.selectedSessionId = sessionId; return copy
            }
            let newGroup = PaneLayoutNode(id: freshId(excluding: &nodeIds), kind: "tabs", sessionIds: [sessionId], selectedSessionId: sessionId)
            return PaneLayoutNode(id: freshId(excluding: &nodeIds), kind: "split", axis: ["left", "right"].contains(placement) ? "horizontal" : "vertical", children: ["left", "top"].contains(placement) ? [newGroup, node] : [node, newGroup])
        }
        return withinLimits(result) ? result : root
    }

    /// Invalid drops are no-ops. In particular, validate the target before
    /// removing the source so stale drag payloads cannot lose a live session.
    public static func moving(root: PaneLayoutNode?, sessionId: String, targetGroupId: String, placement: String, beforeSessionId: String? = nil) -> PaneLayoutNode? {
        guard let root, CoreValidation.identifier(sessionId), ["tab", "left", "right", "top", "bottom"].contains(placement),
              let source = find(root, depth: 0, where: { $0.kind == "tabs" && $0.sessionIds.contains(sessionId) }),
              let target = find(root, depth: 0, where: { $0.kind == "tabs" && $0.id == targetGroupId }),
              beforeSessionId == nil || CoreValidation.identifier(beforeSessionId!) && target.sessionIds.contains(beforeSessionId!) else { return root }
        if source.id == target.id && (source.sessionIds.count == 1 || placement == "tab" && beforeSessionId == sessionId) { return root }
        let ids = allSessionIds(root)
        guard let clean = normalized(root: root, sessionIds: ids), allSessionIds(clean).contains(sessionId),
              find(clean, depth: 0, where: { $0.kind == "tabs" && $0.id == targetGroupId }) != nil,
              var removed = removing(clean, sessionId: sessionId) else { return root }
        var nodeIds = Set(allNodeIds(removed)), inserted = false
        removed = map(removed, depth: 0) { node in
            guard node.kind == "tabs", node.id == targetGroupId else { return node }
            inserted = true
            if placement == "tab" {
                var copy = node
                let index = beforeSessionId.flatMap { copy.sessionIds.firstIndex(of: $0) } ?? copy.sessionIds.count
                copy.sessionIds.insert(sessionId, at: index); copy.selectedSessionId = sessionId
                return copy
            }
            let tab = PaneLayoutNode(id: freshId(excluding: &nodeIds), kind: "tabs", sessionIds: [sessionId], selectedSessionId: sessionId)
            return PaneLayoutNode(id: freshId(excluding: &nodeIds), kind: "split", axis: placement == "left" || placement == "right" ? "horizontal" : "vertical",
                                  children: placement == "left" || placement == "top" ? [tab, node] : [node, tab])
        }
        guard inserted, withinLimits(removed) else { return root }
        return removed
    }

    public static func resizing(root: PaneLayoutNode?, splitId: String, ratio: Double) -> PaneLayoutNode? {
        guard let root, ratio.isFinite else { return root }
        return map(root, depth: 0) { node in
            guard node.kind == "split", node.id == splitId else { return node }
            var copy = node; copy.ratio = clamp(ratio); return copy
        }
    }

    private static func validSessionIds(_ ids: [String]) -> [String] {
        var seen = Set<String>()
        return ids.prefix(maximumSessions).filter { CoreValidation.identifier($0) && seen.insert($0).inserted }
    }
    private static func clamp(_ ratio: Double) -> Double { ratio.isFinite ? min(maximumRatio, max(minimumRatio, ratio)) : 0.5 }
    private static func freshId(excluding used: inout Set<String>) -> String {
        var id: String
        repeat { id = UUID().uuidString } while !used.insert(id).inserted
        return id
    }
    private static func map(_ node: PaneLayoutNode, depth: Int, _ transform: (PaneLayoutNode) -> PaneLayoutNode) -> PaneLayoutNode {
        guard depth < maximumDepth else { return node }
        var copy = node
        if node.kind == "split" { copy.children = node.children.prefix(2).map { map($0, depth: depth + 1, transform) } }
        return transform(copy)
    }
    private static func find(_ node: PaneLayoutNode, depth: Int, where predicate: (PaneLayoutNode) -> Bool) -> PaneLayoutNode? {
        guard depth < maximumDepth else { return nil }
        if predicate(node) { return node }
        for child in node.children.prefix(2) { if let found = find(child, depth: depth + 1, where: predicate) { return found } }
        return nil
    }
    private static func allSessionIds(_ root: PaneLayoutNode) -> [String] {
        var ids: [String] = []
        _ = map(root, depth: 0) { node in if node.kind == "tabs", ids.count < maximumSessions { ids.append(contentsOf: node.sessionIds.prefix(maximumSessions - ids.count)) }; return node }
        return ids
    }
    private static func allNodeIds(_ root: PaneLayoutNode) -> [String] {
        var ids: [String] = []
        _ = map(root, depth: 0) { node in ids.append(node.id); return node }
        return ids
    }
    private static func removing(_ node: PaneLayoutNode, sessionId: String) -> PaneLayoutNode? {
        if node.kind == "tabs" {
            var copy = node; copy.sessionIds.removeAll { $0 == sessionId }
            guard !copy.sessionIds.isEmpty else { return nil }
            if copy.selectedSessionId == sessionId { copy.selectedSessionId = copy.sessionIds[0] }
            return copy
        }
        let children = node.children.compactMap { removing($0, sessionId: sessionId) }
        if children.count == 1 { return children[0] }
        guard children.count == 2 else { return nil }
        var copy = node; copy.children = children; return copy
    }
    private static func withinLimits(_ root: PaneLayoutNode) -> Bool {
        var count = 0
        func visit(_ node: PaneLayoutNode, depth: Int) -> Bool {
            count += 1
            guard count <= maximumNodes, depth < maximumDepth else { return false }
            return node.children.allSatisfy { visit($0, depth: depth + 1) }
        }
        return visit(root, depth: 0)
    }
}

File: native/windows/MightyClaude.Core/PaneLayout.cs
namespace MightyClaude.Core;

/// <summary>A workspace layout consists of tab groups and binary splits.</summary>
public sealed record PaneLayoutNode
{
    public string Id { get; init; } = Wire.Id();
    public string Kind { get; init; } = "tabs";
    public List<string> SessionIds { get; init; } = [];
    public string? SelectedSessionId { get; init; }
    public string? Axis { get; init; }
    public double Ratio { get; init; } = .5;
    public List<PaneLayoutNode> Children { get; init; } = [];
}

public static class PaneLayout
{
    public static IEnumerable<PaneLayoutNode> Groups(PaneLayoutNode? node) => node is null ? [] : node.Kind == "tabs" ? [node] : node.Children.SelectMany(Groups);
    public static IEnumerable<string> Sessions(PaneLayoutNode? node) => Groups(node).SelectMany(g => g.SessionIds);
    private static PaneLayoutNode Tabs(IEnumerable<string> ids, string? active = null) { var values = ids.ToList(); return new() { SessionIds = values, SelectedSessionId = values.Contains(active ?? "") ? active : values.FirstOrDefault() }; }
    private static PaneLayoutNode Split(PaneLayoutNode a, PaneLayoutNode b, string axis, double ratio = .5) => new() { Kind = "split", Axis = axis, Ratio = ratio, Children = [a, b] };
    public static double ClampRatio(double ratio) => double.IsFinite(ratio) ? Math.Clamp(ratio, .15, .85) : .5;

    public static PaneLayoutNode? Normalize(PaneLayoutNode? root, IEnumerable<string> sessionIds, string? active = null)
    {
        var ordered = sessionIds.Take(128).Where(Wire.Identifier).Distinct().ToList();
        if (ordered.Count == 0) return null;
        var allowed = ordered.ToHashSet(); var seen = new HashSet<string>(); var nodes = new HashSet<string>(); var count = 0;
        PaneLayoutNode? Visit(PaneLayoutNode? node, int depth)
        {
            if (node is null || depth >= 16 || ++count > 255) return null;
            var id = Wire.Identifier(node.Id) && nodes.Add(node.Id) ? node.Id : Wire.Id(); nodes.Add(id);
            if (node.Kind == "tabs")
            {
                var ids = (node.SessionIds ?? []).Take(128).Where(id => allowed.Contains(id) && seen.Add(id)).ToList();
                return ids.Count == 0 ? null : new() { Id = id, SessionIds = ids, SelectedSessionId = ids.Contains(node.SelectedSessionId ?? "") ? node.SelectedSessionId : ids[0] };
            }
            if (node.Kind != "split") return null;
            var children = (node.Children ?? []).Take(2).Select(child => Visit(child, depth + 1)).OfType<PaneLayoutNode>().ToList();
            return children.Count switch { 0 => null, 1 => children[0], _ => new() { Id = id, Kind = "split", Axis = node.Axis == "vertical" ? "vertical" : "horizontal", Ratio = ClampRatio(node.Ratio), Children = children } };
        }
        var result = Visit(root, 0);
        var missing = ordered.Where(id => !seen.Contains(id)).ToList();
        if (result is null) result = Tabs(missing, active);
        else if (missing.Count > 0)
        {
            var target = Groups(result).First();
            result = Replace(result, target.Id, target with { SessionIds = target.SessionIds.Concat(missing).ToList() });
        }
        return active is null ? result : Select(result, active);
    }

    public static PaneLayoutNode? Preset(IEnumerable<string> sessionIds, string layout, string? active = null)
    {
        var ids = sessionIds.Take(128).Where(Wire.Identifier).Distinct().ToList(); if (ids.Count == 0) return null;
        if (layout is "focus" or "tabs" or "custom") return Tabs(ids, active);
        PaneLayoutNode Build(IReadOnlyList<string> values, int depth)
        {
            if (values.Count == 1) return Tabs(values);
            var middle = (values.Count + 1) / 2;
            return Split(Build(values.Take(middle).ToArray(), depth + 1), Build(values.Skip(middle).ToArray(), depth + 1), layout == "columns" || depth % 2 == 0 ? "horizontal" : "vertical", (double)middle / values.Count);
        }
        return Build(ids, 0);
    }

    public static PaneLayoutNode Select(PaneLayoutNode root, string sessionId) => root.Kind == "tabs"
        ? root.SessionIds.Contains(sessionId) ? root with { SelectedSessionId = sessionId } : root
        : root with { Children = root.Children.Select(child => Select(child, sessionId)).ToList() };
    public static PaneLayoutNode Resize(PaneLayoutNode root, string splitId, double ratio) => !double.IsFinite(ratio) ? root : root.Id == splitId && root.Kind == "split"
        ? root with { Ratio = ClampRatio(ratio) }
        : root with { Children = root.Children.Select(child => Resize(child, splitId, ratio)).ToList() };
    private static PaneLayoutNode Replace(PaneLayoutNode root, string id, PaneLayoutNode replacement) => root.Id == id ? replacement : root with { Children = root.Children.Select(child => Replace(child, id, replacement)).ToList() };
    private static PaneLayoutNode? Remove(PaneLayoutNode root, string sessionId)
    {
        if (root.Kind == "tabs")
        {
            var ids = root.SessionIds.Where(id => id != sessionId).ToList();
            return ids.Count == 0 ? null : root with { SessionIds = ids, SelectedSessionId = ids.Contains(root.SelectedSessionId ?? "") ? root.SelectedSessionId : ids[0] };
        }
        var children = root.Children.Select(child => Remove(child, sessionId)).OfType<PaneLayoutNode>().ToList();
        return children.Count switch { 0 => null, 1 => children[0], _ => root with { Children = children } };
    }

    /// <summary>Move one existing tab. Index is the insertion slot before removal.</summary>
    public static PaneLayoutNode Move(PaneLayoutNode root, string sessionId, string targetGroupId, string edge = "center", int? index = null)
    {
        var groups = Groups(root).ToList(); var source = groups.FirstOrDefault(g => g.SessionIds.Contains(sessionId)); var target = groups.FirstOrDefault(g => g.Id == targetGroupId);
        if (source is null || target is null || edge is not ("center" or "left" or "right" or "top" or "bottom")) return root;
        if (source.Id == target.Id)
        {
            if (edge == "center")
            {
                var oldIndex = target.SessionIds.IndexOf(sessionId); var slot = Math.Clamp(index ?? target.SessionIds.Count, 0, target.SessionIds.Count); if (oldIndex < slot) slot--;
                var ids = target.SessionIds.Where(id => id != sessionId).ToList(); ids.Insert(Math.Clamp(slot, 0, ids.Count), sessionId);
                return Replace(root, target.Id, target with { SessionIds = ids, SelectedSessionId = sessionId });
            }
            if (source.SessionIds.Count == 1) return root;
        }
        var remaining = Remove(root, sessionId)!;
        target = Groups(remaining).First(g => g.Id == targetGroupId);
        if (edge == "center")
        {
            var ids = target.SessionIds.ToList(); ids.Insert(Math.Clamp(index ?? ids.Count, 0, ids.Count), sessionId);
            return Replace(remaining, target.Id, target with { SessionIds = ids, SelectedSessionId = sessionId });
        }
        var tab = Tabs([sessionId]); var before = edge is "left" or "top";
        var result = Replace(remaining, target.Id, Split(before ? tab : target, before ? target : tab, edge is "left" or "right" ? "horizontal" : "vertical"));
        int Depth(PaneLayoutNode node) => node.Children.Count == 0 ? 0 : 1 + node.Children.Max(Depth);
        int Count(PaneLayoutNode node) => 1 + node.Children.Sum(Count);
        return Depth(result) >= 16 || Count(result) > 255 ? root : result;
    }
}
